Several enhancements

motin opened the pull request:
Work-in-progress PR for ongoing review. 

Fixes #5
Fixes #4
Fixes #9
Fixes #3
Fixes #13
Fixes #6 
Fixes #26

This PR has been rebased on latest master after the merge of https://github.com/mozilla/openwpm-webext-instrumentation/pull/7

motin commented:
@englehardt This is now rebased onto the latest master and can be reviewed. (The code at this point in time is what has made it into v0.6.0 of the JESTr study add-on)

nhnt11 commented on src/lib/http-post-parser.ts:
```diff
@@ -4,9 +4,9 @@ import {
   // WebRequestOnBeforeSendHeadersEventDetails,
 } from "../types/browser-web-request-event-details";
-import { escapeString } from "./string-utils";
+// import { escapeString } from "./string-utils";
```
Observation: Seems like some stuff is getting commented out as a way to save it for later.

motin replied:
Yes indeed. The post parsing code is currently kept in place, but commented, in case it is to be used in the near future.

motin replied:
See https://github.com/mozilla/openwpm-webext-instrumentation/issues/18

nhnt11 commented on src/lib/http-post-parser.ts:
```diff
@@ -205,7 +210,7 @@ export class HttpPostParser {
       Content-Disposition: form-data; name="email"
       [snipped]
-     */
+     * /
```
Hmm, interesting, nested comment blocks :D
I wouldn't be against using single-line-comments for this entire block of "saving this for later" code - seems like it's intended to be uncommented later anyway.

motin replied:
That would indeed be clearer, but "saving for later" code should not even exist :) - This is a very temporary code block (it will either be removed or reinstated very soon). Referenced in the code cleanup issue at https://github.com/mozilla/openwpm-webext-instrumentation/issues/21

motin commented on src/background/http-instrument.ts:
```diff
@@ -296,9 +307,10 @@ export class HttpInstrument {
-            encodingType,
-          );
+          const postObj: ParsedPostRequest = postParser
+            .parsePostRequest
```
With `encodingType` commented, prettier deemed it was the right choice to do so :)

motin replied:
Created issue for establishing style guidelines: https://github.com/mozilla/openwpm-webext-instrumentation/issues/25

motin commented on src/background/http-instrument.ts:
```diff
@@ -231,12 +232,22 @@ export class HttpInstrument {
+    const tab =
+      details.tabId > -1
+        ? await browser.tabs.get(details.tabId)
+        : { windowId: undefined, incognito: undefined, url: undefined };
```
Indeed an artifact of using TypeScript, and using `undefined` here was a lazy choice to not have to wrap all tab-relevant assignments in if-clauses (so that they are truly undefined if the tab object is not available). Added a code cleanup issue to catch this technical debt: https://github.com/mozilla/openwpm-webext-instrumentation/issues/21

motin commented on src/background/cookie-instrument.ts:
```diff
@@ -1,7 +1,43 @@
+    expiryTimeString = expiryTimeDate.toISOString();
+  }
+  javascriptCookie.expiry = expiryTimeString;
+  javascriptCookie.is_http_only = boolToInt(cookie.httpOnly);
```
It is, but the assignments are rarely based on what type the attribute has, rather its name and other circumstances. Being specific here is the conservative choice. Removing boolToInt however would be a good cleanup to consider - add to https://github.com/mozilla/openwpm-webext-instrumentation/issues/21

englehardt commented on src/background/http-instrument.ts:
```diff
@@ -493,10 +504,24 @@ export class HttpInstrument {
 
+    const responseStatus = details.statusCode;
+    const responseStatusText = details.statusLine;
+
```
This listener doesn't currently have enough data to track redirects. The data we have available to us in the [details object](https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/webRequest/onBeforeRedirect#details) is:
* `redirectUrl`: The URL that will be used for the new request
* `url`: the URL requested for the previous request which resulted in a `Location` header response.

We can save these as `old_request_url` and `new_request_url`. Ideally we should file a bug on Bugzilla to get `redirectRequestId` added to the Web Extensions API.

motin replied:
> We can save these as old_request_url and new_request_url.
Done
> Ideally we should file a bug on Bugzilla to get redirectRequestId added to the Web Extensions API.
Agree

englehardt commented on src/background/navigation-instrument.ts:
```diff
@@ -0,0 +1,68 @@
+        tab_opener_tab_id: tab.openerTabId,
+        frame_id: details.frameId,
+        parent_frame_id: (details as any).parent_frame_id, // An undocumented property
+        window_width: window.width,
```
Why collect window size info?

motin replied:
Ultimately, we are exploring if window/tab/frame dimensions are useful heuristics for tracking behavior. See discussion here: https://github.com/mozilla/openwpm-webext-instrumentation/issues/6

motin commented on src/background/http-instrument.ts:
```diff
@@ -231,12 +232,22 @@ export class HttpInstrument {
     const update = {} as HttpRequest;
 
+    update.incognito = boolToInt(tab.incognito);
     update.crawl_id = crawlID;
```
My original idea was also to combine the attributes, but I wouldn't want to remove or change the crawl_id attribute before having this discussion. Main question is whether or not crawl_id is shared between many browser sessions or if it is unique to one browser spawn? If the latter, we should indeed merge the two attributes.

englehardt replied:
It is shared between sessions (it's essentially a "crawler index" id generated by the TaskManager to identify which one of the N browsers it is managing is being interacted with). However that consistency is not a property we rely on and we could instead generate a new crawl_id every time the browser is restarted.

motin replied:
A name of `session_id` is a bit too generic and suggests a relation with an active session id on the currently visited site. It could be called something like `browser_process_uuid`, `browser_process_id`, `browser_launch_uuid`, `browser_launch_id`.

englehardt replied:
Let's go with `browser_launch_id`. Filed https://github.com/mozilla/openwpm-webext-instrumentation/issues/29 for this.

motin commented on src/background/cookie-instrument.ts:
```diff
@@ -1,7 +1,43 @@
 import Cookie = browser.cookies.Cookie;
 import OnChangedCause = browser.cookies.OnChangedCause;
-import { JavascriptCookieChange } from "../schema";
+import { JavascriptCookie, JavascriptCookieRecord } from "../schema";
```
Fully agree. I propose a follow-up bug so that we can also address other relevant schema changes that we may want to do now rather than later.

englehardt replied:
Done https://github.com/mozilla/openwpm-webext-instrumentation/issues/27

motin commented on src/background/cookie-instrument.ts:
```diff
@@ -21,52 +57,34 @@ export class CookieInstrument {
     browser.cookies.onChanged.addListener(this.onChangedListener);
   }
 
+  public async saveAllCookies(crawlID) {
```
Indeed, this was added in relation to https://github.com/motin/jestr-pioneer-shield-study/issues/2. Added it to the instrumentation rather than JESTr so that privacy-related extensions can perform something similar to the cookie table dumping that OpenWPM performs using the profile contents on disk.

englehardt replied:
Cool. Added https://github.com/mozilla/openwpm-webext-instrumentation/issues/28 to remind us to document this at some point.

motin commented:
> One more general question related to the navigation instrument. Is the idea here to record timestamp and URLs that correspond to (tab_id, frame_id, window_id, session_id) key and then attribute the top-level URL to all records with a matching (tab_id, frame_id, window_id, session_id) key which have a timestamp greater than the navigation event for that URL and less than the navigation event for the next URL? If so, have you explored potential timing issues related to that? For example, how do you tag responses which arrive after the navigation event has fired? A possibly related issue: EFForg/privacybadger#1997

Moved this over to https://github.com/mozilla/openwpm-webext-instrumentation/issues/9#issuecomment-440438090

> I see three larger changes:

>     Renaming the JavascriptCookies table

I propose a follow-up issue as per the comment above.
>     Adding extra columns for redirects

Done
>     Renaming and refactoring the session id / crawl id

I propose a follow-up issue.

> I also added a couple questions and nits. @motin Happy to have this merged now as long as we file follow-up issues to address the above three points after the questions / nits are resolved.

Thanks for the review! Several mentioned things have been resolved. Let's merge and file follow-up issues on the rest :)

motin commented:
Added a fix for #26 as well as restored the HttpRequest.referrer property.
